[Upgrade Assistant] [Core] Data streams readonly flow

Bamieh opened the pull request:
## Data streams readonly flow

Implement marking data streams as read only in addition to the already existing reindex flow.

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/274022)
* Commit: 92512b31f833989ed57e191a2c5f4f70c820d845

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/274022#0194dbcf-563e-4c9d-bb61-ff9d949553cc)

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274022","number":274022,"commit":"92512b31f833989ed57e191a2c5f4f70c820d845"}],"number":274022}
buildkite-pr-comment-->

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/274087)
* Commit: 427e179d739dec9f18435579c6b9f68c9cd68614

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/dd8084645e88ad2791c75bdf79f786aefe23ceaa) | [after](https://github.com/elastic/kibana/commit/427e179d739dec9f18435579c6b9f68c9cd68614) | diff |
| --- | --- | --- | --- |
| `upgradeAssistant` | 171 | 176 | +5 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/dd8084645e88ad2791c75bdf79f786aefe23ceaa) | [after](https://github.com/elastic/kibana/commit/427e179d739dec9f18435579c6b9f68c9cd68614) | diff |
| --- | --- | --- | --- |
| `upgradeAssistant` | 189.7KB | 195.7KB | +6.0KB |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/dd8084645e88ad2791c75bdf79f786aefe23ceaa) | [after](https://github.com/elastic/kibana/commit/427e179d739dec9f18435579c6b9f68c9cd68614) | diff |
| --- | --- | --- | --- |
| `upgradeAssistant` | 23.6KB | 23.7KB | +169.0B |



### History
* :broken_heart: [Build #274022](https://buildkite.com/elastic/kibana-pull-request/builds/274022) failed 92512b31f833989ed57e191a2c5f4f70c820d845

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274087","number":274087,"commit":"427e179d739dec9f18435579c6b9f68c9cd68614"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274022","number":274022,"commit":"92512b31f833989ed57e191a2c5f4f70c820d845"}],"number":274087}
buildkite-pr-comment-->

jloleysens commented on x-pack/platform/plugins/private/upgrade_assistant/server/lib/data_streams/data_stream_migration_service.ts:
```diff
@@ -321,5 +328,16 @@ export const dataStreamReindexServiceFactory = ({
+    async readonlyIndices(indices: string[]) {
+      for (const index of indices) {
+        const unfreeze = await esClient.indices.unfreeze({ index });
+        const addBlock = await esClient.indices.addBlock({ index, block: 'read_only' });
```
```suggestion
        const addBlock = await esClient.indices.addBlock({ index, block: 'write' });
```

Based on feedback from ES I think we need to use `write` here... But I may have outdated info

Bamieh replied:
We need to check I took this code from @gsoldevila so we'd need to update it on both ends. Ultimately we'd share this function call between data stream service and normal reindexing service

jloleysens reviewed (commented):
Server-side LGTM, assuming test coverage is incoming. Proceeding to public

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/274249)
* Commit: 764a09120e2d3f25af242bf64d5eb6e231a0edb2

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/dd8084645e88ad2791c75bdf79f786aefe23ceaa) | [after](https://github.com/elastic/kibana/commit/764a09120e2d3f25af242bf64d5eb6e231a0edb2) | diff |
| --- | --- | --- | --- |
| `upgradeAssistant` | 171 | 176 | +5 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/dd8084645e88ad2791c75bdf79f786aefe23ceaa) | [after](https://github.com/elastic/kibana/commit/764a09120e2d3f25af242bf64d5eb6e231a0edb2) | diff |
| --- | --- | --- | --- |
| `upgradeAssistant` | 189.7KB | 195.7KB | +6.0KB |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/dd8084645e88ad2791c75bdf79f786aefe23ceaa) | [after](https://github.com/elastic/kibana/commit/764a09120e2d3f25af242bf64d5eb6e231a0edb2) | diff |
| --- | --- | --- | --- |
| `upgradeAssistant` | 23.6KB | 23.7KB | +169.0B |



### History
* :green_heart: [Build #274087](https://buildkite.com/elastic/kibana-pull-request/builds/274087) succeeded 427e179d739dec9f18435579c6b9f68c9cd68614
* :broken_heart: [Build #274022](https://buildkite.com/elastic/kibana-pull-request/builds/274022) failed 92512b31f833989ed57e191a2c5f4f70c820d845

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274249","number":274249,"commit":"764a09120e2d3f25af242bf64d5eb6e231a0edb2"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274087","number":274087,"commit":"427e179d739dec9f18435579c6b9f68c9cd68614"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274022","number":274022,"commit":"92512b31f833989ed57e191a2c5f4f70c820d845"}],"number":274249}
buildkite-pr-comment-->

jloleysens commented on x-pack/platform/plugins/private/upgrade_assistant/public/application/components/es_deprecations/deprecation_types/data_streams/resolution_table_cell.tsx:
```diff
@@ -16,139 +16,167 @@ import {
+      'xpack.upgradeAssistant.esDeprecations.dataStream.reindexInProgressText',
+      {
+        defaultMessage:
+          '{resolutionType, select, reindex {Reindexing} readonly {Marking as readonly} other {Migration}} in progress…',
```
```suggestion
          '{resolutionType, select, reindex {Reindexing} read only {Marking as read only} other {Migration}} in progress…',
```

nit we must consistently do `readonly` or `read only`. I think `read only` is probably better.

Bamieh replied:
Updated. `readonly` is only used now as the resolution type. the rest is `read only`

jloleysens commented on x-pack/platform/plugins/private/upgrade_assistant/public/application/components/es_deprecations/deprecation_types/data_streams/flyout/steps/details/details_step.tsx:
```diff
@@ -233,17 +234,31 @@ export const DataStreamDetailsFlyoutStep: React.FunctionComponent<{
-                    color={status === DataStreamReindexStatus.cancelled ? 'warning' : 'primary'}
-                    iconType={status === DataStreamReindexStatus.cancelled ? 'play' : undefined}
-                    onClick={startReindex}
+                    color={status === DataStreamMigrationStatus.cancelled ? 'warning' : 'accent'}
```
I think this should just be an unfilled primary color button:

<img width="493" alt="Screenshot 2025-02-07 at 18 10 41" src="https://github.com/user-attachments/assets/5e8046c0-a8e3-4361-b020-650362b39d49" />

jloleysens reviewed (approved):
Public side LGTM.

Tested locally and I was able to set a couple of data streams to read only.

---

One slightly weird situation is that when the "Mark as read only" fails and I get this try again button:

<img width="288" alt="Screenshot 2025-02-07 at 18 00 48" src="https://github.com/user-attachments/assets/53aaf988-8e25-43b8-a857-d49044027063" />

After I click it I get:

<img width="511" alt="Screenshot 2025-02-07 at 18 00 53" src="https://github.com/user-attachments/assets/2e69dde5-9ac6-4640-8d50-30a9aa5759ed" />

I can click "< Back" to get back to the "Reindex" flow if I wanted to was just a little tricky to spot.

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/274397)
* Commit: 811d0f0e435ac63185e98dafcdff33d56c76cd04

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/f2b8944ea845110599a2429495954dcf0f2c7e60) | [after](https://github.com/elastic/kibana/commit/811d0f0e435ac63185e98dafcdff33d56c76cd04) | diff |
| --- | --- | --- | --- |
| `upgradeAssistant` | 171 | 176 | +5 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/f2b8944ea845110599a2429495954dcf0f2c7e60) | [after](https://github.com/elastic/kibana/commit/811d0f0e435ac63185e98dafcdff33d56c76cd04) | diff |
| --- | --- | --- | --- |
| `upgradeAssistant` | 189.7KB | 195.9KB | +6.2KB |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/f2b8944ea845110599a2429495954dcf0f2c7e60) | [after](https://github.com/elastic/kibana/commit/811d0f0e435ac63185e98dafcdff33d56c76cd04) | diff |
| --- | --- | --- | --- |
| `upgradeAssistant` | 23.6KB | 23.7KB | +171.0B |



### History
* :green_heart: [Build #274249](https://buildkite.com/elastic/kibana-pull-request/builds/274249) succeeded 764a09120e2d3f25af242bf64d5eb6e231a0edb2
* :green_heart: [Build #274087](https://buildkite.com/elastic/kibana-pull-request/builds/274087) succeeded 427e179d739dec9f18435579c6b9f68c9cd68614
* :broken_heart: [Build #274022](https://buildkite.com/elastic/kibana-pull-request/builds/274022) failed 92512b31f833989ed57e191a2c5f4f70c820d845

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274397","number":274397,"commit":"811d0f0e435ac63185e98dafcdff33d56c76cd04"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274249","number":274249,"commit":"764a09120e2d3f25af242bf64d5eb6e231a0edb2"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274087","number":274087,"commit":"427e179d739dec9f18435579c6b9f68c9cd68614"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274022","number":274022,"commit":"92512b31f833989ed57e191a2c5f4f70c820d845"}],"number":274397}
buildkite-pr-comment-->

Bamieh commented on x-pack/platform/plugins/private/upgrade_assistant/public/application/components/es_deprecations/deprecation_types/data_streams/readonly_state.tsx:
```diff
@@ -0,0 +1,102 @@
+  }
+
+  let processedCount = 0;
+  const batches = chunk(indicesRequiringUpgrade, batchSize);
```
we cannot do it on the server side without a notion of "state". since we dropped that we have to do it on the client side.
If the user drops in the middle of it that is OK since they can just pick up from where they left off from anytime.
We cannot keep an inmemory state on the server side in case we have multiple nodes. We cannot check the progress status if it is happening on the server side.

Bamieh commented on x-pack/platform/plugins/private/upgrade_assistant/public/application/components/es_deprecations/deprecation_types/data_streams/readonly_state.tsx:
```diff
@@ -0,0 +1,102 @@
+    };
+  }
+
+  return {
```
yes return value is needed to mark the generator as complete

Bamieh commented:
- [x] Fixed bubbling the error message into the failure callout
<img width="491" alt="image" src="https://github.com/user-attachments/assets/37d2945d-d7b8-4613-9ea5-98575687ebd1" />

- [x] Fixed the try again readonly flow
- [x] Fixed batch size to `10`

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/274526)
* Commit: 816da5c15ae1f621055f02e1978eb93df92d15be

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/2046aed6a60a615b6179a4a6b81003f4b91ed7d6) | [after](https://github.com/elastic/kibana/commit/816da5c15ae1f621055f02e1978eb93df92d15be) | diff |
| --- | --- | --- | --- |
| `upgradeAssistant` | 171 | 176 | +5 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/2046aed6a60a615b6179a4a6b81003f4b91ed7d6) | [after](https://github.com/elastic/kibana/commit/816da5c15ae1f621055f02e1978eb93df92d15be) | diff |
| --- | --- | --- | --- |
| `upgradeAssistant` | 189.7KB | 196.5KB | +6.7KB |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/2046aed6a60a615b6179a4a6b81003f4b91ed7d6) | [after](https://github.com/elastic/kibana/commit/816da5c15ae1f621055f02e1978eb93df92d15be) | diff |
| --- | --- | --- | --- |
| `upgradeAssistant` | 23.6KB | 23.7KB | +171.0B |



### History
* :green_heart: [Build #274397](https://buildkite.com/elastic/kibana-pull-request/builds/274397) succeeded 811d0f0e435ac63185e98dafcdff33d56c76cd04
* :green_heart: [Build #274249](https://buildkite.com/elastic/kibana-pull-request/builds/274249) succeeded 764a09120e2d3f25af242bf64d5eb6e231a0edb2
* :green_heart: [Build #274087](https://buildkite.com/elastic/kibana-pull-request/builds/274087) succeeded 427e179d739dec9f18435579c6b9f68c9cd68614
* :broken_heart: [Build #274022](https://buildkite.com/elastic/kibana-pull-request/builds/274022) failed 92512b31f833989ed57e191a2c5f4f70c820d845

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274526","number":274526,"commit":"816da5c15ae1f621055f02e1978eb93df92d15be"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274397","number":274397,"commit":"811d0f0e435ac63185e98dafcdff33d56c76cd04"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274249","number":274249,"commit":"764a09120e2d3f25af242bf64d5eb6e231a0edb2"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274087","number":274087,"commit":"427e179d739dec9f18435579c6b9f68c9cd68614"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/274022","number":274022,"commit":"92512b31f833989ed57e191a2c5f4f70c820d845"}],"number":274526}
buildkite-pr-comment-->

Bamieh commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.18|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/210302">](https://github.com/elastic/kibana/pull/210302)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.6.4 -->
